frontend/src/app/sessions/user-sessions-wrapper/create-session/create-readonly-session/create-readonly-session.component.ts

```diff
@@ -60,33 +62,82 @@ export class CreateReadonlySessionComponent implements OnInit {
   ) {}
 
   ngOnInit(): void {
-    this.modelService.models$
-      .pipe(untilDestroyed(this), filter(Boolean))
-      .subscribe((models) => (this.models = models));
-
     this.projectService.project$
       .pipe(untilDestroyed(this), filter(Boolean))
       .subscribe((project) => (this.projectSlug = project.slug));
 
+    this.loadToolsAndModels().subscribe(([models, allVersions]) => {
+      this.resolveVersionCompatibility(models, allVersions);
+    });
+
     this.userSessionService.loadSessions();
-    this.toolService.getTools().subscribe();
+  }
+
+  loadToolsAndModels(): Observable<[Model[], ToolVersionWithTool[]]> {
+    return combineLatest([
+      this.modelService.models$.pipe(untilDestroyed(this), filter(Boolean)),
+      this.toolService.getVersionsForTools(),
+    ]).pipe(
+      tap(([_, versions]) => {
+        this.allToolVersions = versions;
+      }),
+    );
+  }
+
+  resolveVersionCompatibility(
+    models: Model[],
+    allVersions: ToolVersionWithTool[],
+  ): void {
+    this.models = [];
+
+    const modelsWithVersion = models.filter((m) => m.version);
+    for (const model of modelsWithVersion) {
+      const extendedModel = model as ModelWithCompatibility;
+      extendedModel.compatibleVersions = [
+        this.findVersionByID(model.version!.id, allVersions)!,
+      ];
+
+      for (const version of allVersions!) {
+        if (version.config.compatible_versions.includes(model.version!.id)) {
+          extendedModel.compatibleVersions.push(version);
+        }
+      }
+
+      this.models?.push(extendedModel);
+    }
+  }
+
+  findVersionByID(id: number, allVersions: ToolVersionWithTool[]) {
+    return allVersions.find((v) => v.id === id);
   }
 
   get tools(): Tool[] | undefined {
-    if (this.models === undefined || this.toolService.tools === undefined) {
+    if (this.models === undefined) {
       return undefined;
     }
-    const toolIds = this.models.map((m) => m.tool.id);
-    return this.toolService.tools.filter((t) => toolIds?.includes(t.id));
+    return this.models
+      ?.map((m) => m.compatibleVersions.map((version) => version.tool))
+      .flat()
+      .filter(
+        (value, index, self) =>
+          index === self.findIndex((tool) => tool.id === value.id),
+      );
```

Comment:
I don't really understand what the `filter` does here. So it would be great to either do it differently here so that it is directly understandable, outsource it to a function with a proper descriptive name (if that is possible?), or add a comment here explaining what is being done.

Reply:
It removes duplicates from the list. I've outsourced it to a function `removesToolDuplicates`.